### PR TITLE
BFloat16: enable prepacked weights's inference

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48994 enable mkldnn conv2d backward to support mkldnn tensor input
* **#48922 BFloat16: enable prepacked weights's inference**

Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin , the error is caused by runing old cpu device(only support avx2), but onednn need the device support avx512 at least for bfloat16 path, so I want to add a gloabal context like **torch._C.has_mkldnn** to skip bfloat16 test case if cpu device not support avx512,. is it ok for your side?

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, There has another option: fall back fp32 path in ideep if onednn is not supported on some old device, this option don't need to change pytorch code except updata ideep. if this option is ok, I will updata ideep at first. Thanks!

**Comment by jgong5** [comment]
> There has another option: fall back fp32 path in ideep if onednn is not supported on some old device, this option don't need to change pytorch code except updata ideep. if this option is ok, I will updata ideep at first. Thanks!

@VitalyFedyunin This option sounds better to us since it does not require extra CPU arch check from the PyTorch side. IDEEP will support BF16 regardless of CPU archs. We will have a quick upgrade on IDEEP if it looks good to you. :-)

**Review by VitalyFedyunin — DISMISSED** [review]

**Comment by VitalyFedyunin** [comment]
Current error doesn't tell anything about avx support:

```
        x_mkldnn = x if x.is_mkldnn else x.to_mkldnn()
        y_mkldnn = torch._C._nn.mkldnn_linear(x_mkldnn, self.weight, self.bias)
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
        y = y_mkldnn if x.is_mkldnn else y_mkldnn.to_dense()
```

Falling back to fp32 path will create invalid impression for users. It is much better to error out with nice warning when something is not supported. 

For tests you don't need to introduce new global context. Something simple like @skipIf(test_arch(),"not supported architecture") will suffice.

**Comment by jgong5** [comment]
> Falling back to fp32 path will create invalid impression for users. It is much better to error out with nice warning when something is not supported.

@VitalyFedyunin Thanks for the suggestions. So, we will also add cpu arch check inside `use_mkldnn` (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L227) so that it does not support bf16 without avx512. Does that make sense to you?

**Comment by VitalyFedyunin** [comment]
Ok we fixed tests, but can we please make sure that we are going to have nice error if someone tries it on old CPU

**Comment by XiaobingSuper** [comment]
> Ok we fixed tests, but can we please make sure that we are going to have nice error if someone tries it on old CPU

There will has a simple API to detect cpu device at next onednn upgrade,, and will report error if one cpu not support bfloat16.

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin , please help review it. Thanks!

**Review by VitalyFedyunin — APPROVED** [review]

File: test/test_mkldnn.py [hunk 1]
```diff
@@ -192,6 +203,30 @@ def test_conv2d(self):
             self._test_serialization(mkldnn_conv2d, (x.to_mkldnn(),))
             self._test_tracing(mkldnn_conv2d, (x.to_mkldnn(),))
 
+    @unittest.skipIf(not has_bf16_support(), "OneDNN bfloat16 path requires AVX512")
```
Comment (ngimel) [inline test/test_mkldnn.py:206]: these tests should not be skipped if there's no bf16 support, instead they should be catching and asserting that appropriate errors are raised.

File: torch/utils/mkldnn.py [hunk 2]
```diff
@@ -178,25 +178,25 @@ def forward(self, x):
...
 
 
-def to_mkldnn(module):
-    def m_fn(m):
+def to_mkldnn(module, dtype=torch.float):
+    def m_fn(m, d):
         if isinstance(m, torch.nn.Linear):
-            return MkldnnLinear(m)
+            return MkldnnLinear(m, d)
         elif isinstance(m, torch.nn.Conv1d):
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/utils/mkldnn.py:186]: Dtype argument is added on Conv1d, Conv3d and will assert if given dtype is not supported.

File: test/test_mkldnn.py [hunk 3]
```diff
@@ -19,6 +19,17 @@
 
 from torch.autograd.gradcheck import gradgradcheck, gradcheck
 
+def has_bf16_support():
+    import subprocess
+    try:
+        # for bf16 path, OneDNN requires the cpu has intel avx512 with avx512bw,
+        # avx512vl, and avx512dq.
+        cmd = "grep avx512bw /proc/cpuinfo | grep avx512vl | grep avx512dq"
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply test/test_mkldnn.py:28]: For windows, bf16 cases are skiped now, the reason is that I can't find a easy way to check the device as linux.

File: test/test_mkldnn.py (hunk 1 above)
Comment (XiaobingSuper) [reply test/test_mkldnn.py]: Add test cases if there's no bf16 support.

File: torch/utils/mkldnn.py [hunk 4]
```diff
@@ -178,25 +177,28 @@ def forward(self, x):
...
-            return MkldnnConv1d(m)
+            return MkldnnConv1d(m, d)
         elif isinstance(m, torch.nn.Conv2d):
-            return MkldnnConv2d(m)
+            return MkldnnConv2d(m, d)
         elif isinstance(m, torch.nn.Conv3d):
-            return MkldnnConv3d(m)
+            return MkldnnConv3d(m, d)
         elif isinstance(m, torch.nn.BatchNorm2d) or isinstance(m, torch.nn.BatchNorm3d):
             return MkldnnBatchNorm(m)
```
Comment (ngimel) [inline torch/utils/mkldnn.py:194]: why is dtype ignored for batchnorm? Does it expect its parameters to remain in fp32?

**Review by ngimel — COMMENTED** [review]
This PR would really benefit from better description and documentation, e.g. the fact that bfloat16 in mkldnn is supported only on avx512 processors really should be put in the description and in the Note somewhere in the code. 
Similarly, I would like to understand why bias is never converted to bfloat16. If it is by design, it should be commented upon and put in a note. 
Similarly, for BatchNorm parameters - if OneDNN wants them in fp32, that's fine, but there should be a comment about this.

File: torch/utils/mkldnn.py (hunk 4 above)
Comment (XiaobingSuper) [reply torch/utils/mkldnn.py:194]: Yes, batchnorm's parameters remain in fp32.

File: torch/utils/mkldnn.py [hunk 5]
```diff
@@ -1,11 +1,10 @@
...
 
 
 class MkldnnLinear(torch.jit.ScriptModule):
-    def __init__(self, dense_module):
+    def __init__(self, dense_module, dtype):
         super(MkldnnLinear, self).__init__()
-        self.register_buffer('weight', dense_module.weight.to_mkldnn())
+        self.register_buffer('weight', dense_module.weight.to_mkldnn(dtype))
         if dense_module.bias is not None:
             self.register_buffer('bias', dense_module.bias.to_mkldnn())
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/utils/mkldnn.py:9]: Yes, for better accuracy, bias is expected fp32 type fro bfloat16 module.

File: test/test_mkldnn.py [hunk 6]
```diff
@@ -144,53 +156,103 @@ def test_repr(self):
...
             self.assertEqual(y_aten, y_mkldnn)
 
-            self._test_serialization(mkldnn_conv1d, (x.to_mkldnn(),))
-            self._test_tracing(mkldnn_conv1d, (x.to_mkldnn(),))
+            self._test_serialization(mkldnn_conv, (input.to_mkldnn(),))
+            self._test_tracing(mkldnn_conv, (input.to_mkldnn(),))
+
+    def test_conv1d(self):
+        for groups in [1, 4]:
+            N = torch.randint(3, 10, (1,)).item()
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply test/test_mkldnn.py:183]: Changed.

**Comment by XiaobingSuper** [comment]
> This PR would really benefit from better description and documentation, e.g. the fact that bfloat16 in mkldnn is supported only on avx512 processors really should be put in the description and in the Note somewhere in the code.
> Similarly, I would like to understand why bias is never converted to bfloat16. If it is by design, it should be commented upon and put in a note.
> Similarly, for BatchNorm parameters - if OneDNN wants them in fp32, that's fine, but there should be a comment about this.

Yes, I add some comments in **test_mkldnn.py** and **torch/utils/mkldnn.py** to describe them.

File: test/test_mkldnn.py [hunk 7]
```diff
@@ -250,24 +289,98 @@ def test_relu_(self):
...
+        else:
+            msg = "mkldnn_relu: bf16 path needs the cpu support avx512bw, avx512vl and avx512dq"
+            self.assertRaisesRegex(RuntimeError,
+                                   msg,
+                                   lambda: torch.relu(x_bf16.to_mkldnn()))
+
+    @unittest.skipIf(IS_WINDOWS, "Limit support for bf16 path")
+    def test_relu_inplace_bf16(self):
+        x = torch.randn((4, 5), dtype=torch.float32) * 10
+        x_bf16 = x.clone().bfloat16()
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply test/test_mkldnn.py]: Changed.

File: test/test_mkldnn.py [hunk 8]
```diff
@@ -250,24 +289,98 @@ def test_relu_(self):
...
+            y_bf16 = torch.relu(x_bf16.to_mkldnn()).to_dense(torch.float32)
+            self.assertEqual(y, y_bf16, atol=1e-1, rtol=1e-3)
+        else:
+            msg = "mkldnn_relu: bf16 path needs the cpu support avx512bw, avx512vl and avx512dq"
+            self.assertRaisesRegex(RuntimeError,
+                                   msg,
+                                   lambda: torch.relu(x_bf16.to_mkldnn()))
+
+    @unittest.skipIf(IS_WINDOWS, "Limit support for bf16 path")
+    def test_relu_inplace_bf16(self):
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply test/test_mkldnn.py:292]: Thanks, I changed it.

**Comment by XiaobingSuper** [comment]
@ngimel , the failed case seems not related to this PR.

**Review by ngimel — APPROVED** [review]
